packages/deployer/src/compiler.ts

```diff
@@ -44,123 +49,59 @@ const SOLC_BIN_DIR = path.join(__dirname, '..', '..', 'solc_bin');
-                }
-            }
-        }
-        return sources;
-    }
     /**
      * Instantiates a new instance of the Compiler class.
      * @param opts Options specifying directories, network, and optimization settings.
      * @return An instance of the Compiler class.
      */
     constructor(opts: CompilerOptions) {
-        this._contractDirs = opts.contractDirs;
+        this._contractsDir = opts.contractsDir;
         this._networkId = opts.networkId;
         this._optimizerEnabled = opts.optimizerEnabled;
         this._artifactsDir = opts.artifactsDir;
         this._specifiedContracts = opts.specifiedContracts;
+        this._nameResolver = new NameResolver(path.resolve(this._contractsDir));
+        const resolver = new FallthroughResolver();
+        resolver.appendResolver(new URLResolver());
+        resolver.appendResolver(new NPMResolver(path.resolve('')));
+        resolver.appendResolver(new FSResolver());
+        resolver.appendResolver(this._nameResolver);
+        this._resolver = resolver;
     }
     /**
      * Compiles selected Solidity files found in `contractsDir` and writes JSON artifacts to `artifactsDir`.
      */
     public async compileAsync(): Promise<void> {
         await createDirIfDoesNotExistAsync(this._artifactsDir);
         await createDirIfDoesNotExistAsync(SOLC_BIN_DIR);
-        this._contractSources = {};
-        const contractIdToSourceFileId: ContractIdToSourceFileId = {};
-        const contractDirs = Array.from(this._contractDirs.values());
-        for (const contractDir of contractDirs) {
-            const sources = await Compiler._getContractSourcesAsync(contractDir.path, contractDir.path);
-            _.forIn(sources, (source, sourceFilePath) => {
-                const sourceFileId = constructUniqueSourceFileId(contractDir.namespace, sourceFilePath);
-                // Record the file's source and data
-                if (!_.isUndefined(this._contractSources[sourceFileId])) {
-                    throw new Error(`Found duplicate source files with ID '${sourceFileId}'`);
-                }
-                this._contractSources[sourceFileId] = source;
-                // Create a mapping between the contract id and its source file id
-                const contractId = constructContractId(contractDir.namespace, sourceFilePath);
-                if (!_.isUndefined(contractIdToSourceFileId[contractId])) {
-                    throw new Error(`Found duplicate contract with ID '${contractId}'`);
-                }
-                contractIdToSourceFileId[contractId] = sourceFileId;
-            });
+        let contractNamesToCompile: string[] = [];
+        if (this._specifiedContracts.has(ALL_CONTRACTS_IDENTIFIER)) {
+            const allContracts = this._nameResolver.getAllContracts();
+            contractNamesToCompile = _.map(allContracts, contractSource =>
+                path.basename(contractSource.path, constants.SOLIDITY_FILE_EXTENSION),
```

Comment:
This should never happen. If we have two contracts with the same name - it's hard to write that correctly to artifacts. Truffle just writes it all to single artifact and overrides it. So we end up with the last contract saved and the first lost. I think - we should just throw in that case.

BTW, if you have the collision in deps - that's ok. You can be like:
```
import {A as B} from '...';

contract A extends B {
...
}
```

It will correctly compile A and emit the artifacts. You shouldn't have the collision between the compiation targets though.